[FRONTEND] Add product header in product page

MehdiSaffar opened the pull request:
## Type of change

- [x] New feature (non-breaking change which adds functionality)

# Description

- Product header in product page
- Show information such as title, description, price (and discount), review (rating + rating count), image gallery for both main and detail pictures
- Allows to visualize each picture separately, with custom gallery and pagination at the bottom of the current picture
- Strikethrough for discounted prices
- Has an add to card and add to list
- This PR also has a ChooseListModal implementation
- Quite responsive to width changes (using CSS grid)
- The custom image gallery shows a vertical scroll when overflowing max height of the header

# Screenshots

## Wide page
![image](https://user-images.githubusercontent.com/13124263/101471738-b61a1500-3958-11eb-9afb-99468d23a187.png)
## Narrow page
![image](https://user-images.githubusercontent.com/13124263/101471781-c3cf9a80-3958-11eb-8d44-9867ea28a997.png)
## When loading shows skeleton of all components
Also has an "sweeping" effect to it.
![image](https://user-images.githubusercontent.com/13124263/101471843-e06bd280-3958-11eb-81c6-e99b5ea6a3b2.png)
## Choose list modal
![image](https://user-images.githubusercontent.com/13124263/101471922-f7aac000-3958-11eb-9fb0-4569a379fe16.png)


# Test results

Manual tests.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have checked my code and corrected any misspellings
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

# Potential problems/Important remarks

- Very draft, not sure of the backend object structure
- Needs stricter code in some areas
- Need code that checks whether user is customer or not.

MehdiSaffar committed:
multiple changes, read description

- add choose list modal for choosing a list modal
- add product page header basically working but needs a lot of polishing
- add skeleton for when the product is still loading
- add an error notification when failing to fetch product
- remove extraneous sleep in useFetch
- make the main image appear with the detail images in the array
- add pagination
- give main image fixed height so that pagination doesnt change vertical
position (would be annoying otherwise)
- fixed image problems between chrome and firefox
- positioned the header within the product page
- various other bug fixes and improvements

ozdenizdolu commented on ecommerce/frontend/package.json:
```diff
@@ -9,6 +9,7 @@
         "@testing-library/user-event": "^12.2.0",
         "antd": "^4.8.0",
         "axios": "^0.21.0",
+        "classnames": "^2.2.6",
```
I have merged a `package.json` that contains classnames before, why is it showing it as new?

MehdiSaffar replied:
I might have branched out of master before you

abdullahyldz commented on ecommerce/frontend/src/components/ChooseListModal.js:
```diff
@@ -0,0 +1,54 @@
+            {lists.map((list, ix) => {
+                return (
+                    <List.Item className={'choose-list-modal-list'} onClick={onListClick(list)} key={list.id ?? ix}>
+                        {list.name ?? `Example list name ${ix}`}
```
You can add heart button next to each list item in the popup in case customer wants to add that product to multiple wishlists.
`<Button onClick={onAddToListClicked(ix)} size="large" icon={<HeartOutlined />} />`

MehdiSaffar replied:
Okay it seems that both you and @ozdenizdolu want the user be able to add to multiple lists at once.. guess I'll do it.

ozdenizdolu commented on ecommerce/frontend/src/components/ChooseListModal.js:
```diff
@@ -0,0 +1,54 @@
+            {lists.map((list, ix) => {
+                return (
+                    <List.Item className={'choose-list-modal-list'} onClick={onListClick(list)} key={list.id ?? ix}>
+                        {list.name ?? `Example list name ${ix}`}
```
Maybe instead of selecting a single list, we can have a checkbox display so that user can add the product into multiple lists at the same request, not a must but it's a design choice.

MehdiSaffar replied:
We should discuss this, but usual use case is to add to single list, like in Facebook for example for saved links/videos. Also people usually have a list for an event such as "Birthday" or "Black friday" or "Christmas" things like that.

ozdenizdolu commented on ecommerce/frontend/src/components/product/ProductHeader.js:
```diff
@@ -0,0 +1,92 @@
+export const ProductHeader = ({ product, loading = false }) => {
+    const [isChooseListModalVisible, setIsChooseListModalVisible] = useState(false)
+
+    const onAddToCartClicked = () => {
+        console.log('add to cart', product)
```
This method can be implemented since it's already using a working endpoint.

MehdiSaffar replied:
Oh right, I will do that then.

ozdenizdolu commented on ecommerce/frontend/src/components/product/ProductHeader.js:
```diff
@@ -0,0 +1,92 @@
+                footer={null}
+                title="Choose a list"
+                destroyOnClose>
+                <ChooseListModal onChooseList={onChooseList} />
```
Nitpick: ChooseListModal can be renamed as ChooseListModalInner as it is actually referring to the inner part of modal, not the whole. I mean it does not render a modal by itself.

MehdiSaffar replied:
You are correct I will change the name.

ozdenizdolu commented on ecommerce/frontend/src/components/product/ProductImageGallery.js:
```diff
@@ -0,0 +1,80 @@
+        )
+    }
+
+    console.log(images.length, active)
```
There are many console.logs here and there, make sure you delete most of the unnecessary ones

MehdiSaffar replied:
Yup, will do. Thanks

ozdenizdolu commented on ecommerce/frontend/src/utils.js:
```diff
@@ -105,3 +105,14 @@ export function formatProduct({
+ * Uses correction to offset floating-point inaccuracies.
+ * Works symmetrically for positive and negative numbers.
+ */
+export function round(num, decimalPlaces = 2) {
```
Really useful for prices all over the frontend. I think you should tell everyone on the team that this function exists in `utils.js`

MehdiSaffar replied:
Sure I will announce it.

abdullahyldz commented on ecommerce/frontend/src/components/product/ProductExtra.less:
```diff
@@ -0,0 +1,26 @@
+        font-size: 2.5em;
+        .product-extra-old-price {
+            font-weight: lighter;
+            text-decoration: line-through;
```
I think, the discounted price is hard to read. Black line on top of black font color. We can try out new colors. 

`text-decoration: line-through #E2BE5A; `  is better but not the best

MehdiSaffar replied:
I'll try that one, thanks

abdullahyldz reviewed (approved):
Great work.

ozdenizdolu reviewed (approved):
Most of my requests are optional. I leave them to you. Great work, it looks really nice.

MehdiSaffar committed:
delete extra console log

MehdiSaffar committed:
make add shopping cart button work

MehdiSaffar committed:
line-through color to champagne

MehdiSaffar committed:
add heart button that is filled or empty

MehdiSaffar committed:
add draft code for add and remove product from list

MehdiSaffar committed:
remove cursor:pointer on list item

MehdiSaffar committed:
add loading heart button when adding or removing

MehdiSaffar committed:
add loading skeleton for list modal

MehdiSaffar committed:
fix key, show entire loading only on mount

MehdiSaffar committed:
add error notification for get lists

MehdiSaffar committed:
hide cancel button and handle close properly

MehdiSaffar committed:
show error when product can't be found

MehdiSaffar committed:
remove extra prop in ChooseListModalInner

MehdiSaffar committed:
pass product prop, fix notif msg

MehdiSaffar committed:
remove unused callback

MehdiSaffar commented:
Integrated all requested changes. 

Also added an error page when the product cannot be found. Maybe it should be moved up to ProductPage instead.

![image](https://user-images.githubusercontent.com/13124263/101785959-6c732b00-3b0e-11eb-9ff0-0c6a83b343a8.png)

MehdiSaffar committed:
clarified a couple things with cmts

ozdenizdolu commented:
Here are some comments:

- Height of the images sometimes (which is set to 500px right now) pushes the pagination 1-2 px below and bottom part of pages disappear. (setting it 495px solves it)
- box-shadow: 5px 5px 0px 0px #472836; gives the discounted price box a nice look (adds depth)

ozdenizdolu commented on ecommerce/frontend/src/components/ChooseListModalInner.js:
```diff
@@ -0,0 +1,118 @@
+        try {
+            setLoadingList(list)
+            await sleep(1000) // actual code here
+            notification.success({ description: 'Removed product from list' })
```
I agree with you on not showing the success notification (since heart icon gets filled)

ozdenizdolu replied:
I agree with you on not showing the success notification (since heart icon gets filled)

ozdenizdolu commented on ecommerce/frontend/src/components/ChooseListModalInner.js:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    return (
+        <List loading={loading} size="large" itemLayout="horizontal">
```
This `loading`  prop is not needed as we have discussed.

ozdenizdolu reviewed (approved):
looks great, here are some final comments:

- Height of the images sometimes (which is set to 500px right now) pushes the pagination 1-2 px below and bottom part of pages disappear. (setting it 495px solves it)
- box-shadow: 5px 5px 0px 0px #472836; gives the discounted price box a nice look (adds depth)
- The vendor rating is a little bit ambiguous(the fact that it's a rating), this css style works better as we have discussed on Google Meet:
```css
element.style {
    background: darkgoldenrod;
    display: inline-block;
    padding: 0px 4px;
    color: white;
    font-weight: bold;
    border-radius: 6px;
}
```

MehdiSaffar committed:
remove redundant success notifs

MehdiSaffar committed:
remove loading prop in ChooseListModalInner

MehdiSaffar committed:
smaller main image height

MehdiSaffar committed:
add box shadow to new price

MehdiSaffar committed:
add vendor rating color

MehdiSaffar commented:
Integrated all requested changes. Tomorrow meeting with @furkannane to integrate with backend.

furkannane committed:
Changed product/

Changed the branch accrdingly

furkannane committed:
Minor change

MehdiSaffar committed:
revert env variable change in setings.py

MehdiSaffar committed:
Merge branch 'Backend]-Product/{int}' into F-product-page-header

MehdiSaffar committed:
integrate with backend

- use images from backend
- use short_description instead of description
- round the vendor_rating

furkannane committed:
Added old price

furkannane committed:
Merge branch 'Backend]-Product/{int}' of https://github.com/bounswe/bounswe2020group2 into Backend]-Product/{int}

furkannane committed:
Format is changed according to requests from frontend team

MehdiSaffar committed:
Merge branch 'Backend]-Product/{int}' into F-product-page-header

MehdiSaffar committed:
vendor is object now, fix old_price

MehdiSaffar committed:
Merge branch 'master' into F-product-page-header
